src/strategies/strategies/nato-uniswap-v3/index.ts

```diff
@@ -0,0 +1,133 @@
+
+const POSITION_MANAGER_ABI = [
+  'function balanceOf(address owner) view returns (uint256)',
+  'function tokenOfOwnerByIndex(address owner, uint256 index) view returns (uint256)',
+  'function positions(uint256 tokenId) view returns (uint96 nonce, address operator, address token0, address token1, uint24 fee, int24 tickLower, int24 tickUpper, uint128 liquidity, uint256 feeGrowthInside0LastX128, uint256 feeGrowthInside1LastX128, uint128 tokensOwed0, uint128 tokensOwed1)'
+];
+
+const POOL_ABI = [
+  'function slot0() view returns (uint160 sqrtPriceX96, int24 tick, uint16 observationIndex, uint16 observationCardinality, uint16 observationCardinalityNext, uint8 feeProtocol, bool unlocked)'
+];
+
+function getAmount1ForLiquidity(
+  sqrtLower: number,
+  sqrtUpper: number,
+  sqrtPriceCurrent: bigint,
+  liquidity: bigint
+): bigint {
+  const lower = BigInt(sqrtLower);
+  const upper = BigInt(sqrtUpper);
+  const current = BigInt(sqrtPriceCurrent);
+  const liq = BigInt(liquidity);
+
+  if (current <= lower) return 0n;
+  if (current < upper) {
+    return (liq * (current - lower)) / (2n ** 96n);
+  }
+  return (liq * (upper - lower)) / (2n ** 96n);
+}
+
+async function multicall(
+  provider: JsonRpcProvider,
+  abi: string[],
+  calls: [string, string, any[]][]
+): Promise<any[]> {
+  const iface = new Interface(abi);
+  const results: any[] = [];
+
+  for (const [address, method, params] of calls) {
+    const contract = new Contract(address, abi, provider);
+    try {
+      const result = await contract[method](...params);
+      results.push([[...(Array.isArray(result) ? result : [result])]]);
+    } catch (err: any) {
+      console.warn(`Multicall failed for ${method} on ${address}:`, err?.message);
+      results.push([[undefined]]);
+    }
+  }
+
+  return results;
+}
+
+export const strategy = async (
+  _space: string,
+  _network: string,
+  _provider: any,
+  addresses: string[],
+  options: any,
+  snapshot: number | 'latest'
+) => {
+  const results: Record<string, number> = {};
+
+const provider = new JsonRpcProvider(
+  'https://developer-access-mainnet.base.org',
+  {
+    name: 'base',
+    chainId: 8453
+  }
+);
+
+
+  const [[slot0Result]] = await multicall(provider, POOL_ABI, [
+    [options.poolAddress, 'slot0', []]
+  ]);
+
+  if (!slot0Result || !slot0Result[0]) {
+    throw new Error('❌ Failed to fetch slot0() from the pool contract.');
+  }
+
+  const sqrtPriceX96 = BigInt(slot0Result[0]);
+
+  for (const address of addresses) {
+    const [[balanceResult]] = await multicall(provider, POSITION_MANAGER_ABI, [
+      [options.positionManager, 'balanceOf', [address]]
+    ]);
+
+    const balance = Number(balanceResult);
+    if (!balance) {
+      results[getAddress(address)] = 0;
+      continue;
+    }
+
+    const tokenIdsCall: [string, string, any[]][] = [];
+    for (let i = 0; i < balance; i++) {
+      tokenIdsCall.push([options.positionManager, 'tokenOfOwnerByIndex', [address, i]]);
+    }
```

Comment:
This won't scale well if this strategy is used with overiding strategies if there are thousands of `addresses`, instead send a call per address like in erc20-balance-of strategy